vyper/venom/passes/load_elimination.py

```diff
@@ -0,0 +1,74 @@
+from vyper.venom.analysis import DFGAnalysis, LivenessAnalysis, VarEquivalenceAnalysis
+from vyper.venom.effects import Effects
+from vyper.venom.passes.base_pass import IRPass
+
+
+class LoadElimination(IRPass):
+    """
+    Eliminate sloads, mloads and tloads
+    """
+
+    # should this be renamed to EffectsElimination?
+
+    def run_pass(self):
+        self.equivalence = self.analyses_cache.request_analysis(VarEquivalenceAnalysis)
+
+        for bb in self.function.get_basic_blocks():
+            self._process_bb(bb)
+
+        self.analyses_cache.invalidate_analysis(LivenessAnalysis)
+        self.analyses_cache.invalidate_analysis(DFGAnalysis)
+
+    def equivalent(self, op1, op2):
+        return op1 == op2 or self.equivalence.equivalent(op1, op2)
+
+    def _process_bb(self, bb):
+        transient = ()
+        storage = ()
+        memory = ()
+
+        for inst in bb.instructions:
+            if Effects.MEMORY in inst.get_write_effects():
+                memory = ()
+            if Effects.STORAGE in inst.get_write_effects():
+                storage = ()
+            if Effects.TRANSIENT in inst.get_write_effects():
+                transient = ()
+
+            if inst.opcode == "mstore":
+                # mstore [val, ptr]
+                memory = (inst.operands[1], inst.operands[0])
```

Comment:
Could we just switch it so the order in which the value and ptr is store we cannot always switch them (just to make it simple to reason about it)

Reply:
i rewrote it a bit so that we don't have to look at the reversal of the indexes
https://github.com/vyperlang/vyper/pull/4265/commits/6faffbaa781d41c650d42ac837416acf86868cef